README.md
# Intent Classification Service

- [Intent Classification Service](#intent-classification-service)
  - [Challenge Task](#challenge-task)
  - [Approach](#approach)
    - [Data Preparation](#data-preparation)
    - [Data](#data)
    - [Modeling](#modeling)
      - [Model](#model)
      - [Parameters](#parameters)
- [Instructions](#instructions)
  - [Setup](#setup)
  - [Building or downloading docker image](#building-or-downloading-docker-image)
    - [Building docker image](#building-docker-image)
    - [Dowload pre-built docker image](#dowload-pre-built-docker-image)
    - [Limitation](#limitation)
- [API Documentation](#api-documentation)
- [Data Preparation and Model Training](#data-preparation-and-model-training)
- [Evaluation](#evaluation)
  - [Accuracy](#accuracy)
  - [AU-ROC](#au-roc)
    - [Overall](#overall)
    - [Per Intent Class](#per-intent-class)
  - [Classificaiton Report](#classificaiton-report)
- [Future Improvements](#future-improvements)


## Challenge Task

Your task is to implement a neural network-based intent classifier that can be used to provide inferencing service via an HTTP Service. The boiler plate for the Service is implemented in file `server.py` and you'll have to implement the API function for inferencing as per the API documentation provided below. The neural network interface has been defined in `intent_classifer.py`. You can add any methods and functionality to this class you deem necessary.

You may use any deep learning library (Tensorflow, Keras, PyTorch, ...) you wish and you can also use pre-existing components for building the network architecture if these would be useful in real-life production systems. Provide tooling and instructions for training the network from scratch.

Also provide a jupyter notebook for model development which trains and tests the model. The final output of this notebook should be trained models with respect to their tests. Evaluation metrics should include the following:

1. Accuracy
2. Precision
3. Recall
4. F1
5. Any other metric that you think is suitable for the comparison

In addition, the same notebook provides a section to evaluate models in production. Assuming the following scenario:

You have both of the models in production and no labeled data is available to you. How would you compare them? Which metrics would you use for this kind of comparison? For example, you can use metrics based on confidence values or related ones.

## Approach

In this notebook we are training a multilingual intent classification, for the purpose of POC I am selecting following languages:

1. English
2. Hindi
3. Spanish

### Data Preparation

The given ATIS dataset is provided in English, I have created a parallel dataset using google translation for Hindi and Spanish.

In the training dataset, we have the following distribution of data:

```
 flight                        3426
 airfare                        403
 ground_service                 235
 airline                        148
 abbreviation                   108
 aircraft                        78
 flight_time                     52
 quantity                        49
 distance                        20
 city                            18
 airport                         18
 ground_fare                     17
 flight+airfare                  17
 capacity                        16
 flight_no                       12
 meal                             6
 restriction                      5
 airline+flight_no                2
 ground_service+ground_fare       1
 airfare+flight_time              1
 cheapest                         1
 aircraft+flight+flight_no        1
```

From what I can observe, the data is quite unbalanced and also some of the classes seems to be a combination of others, which gives me an indication that we can perhaps model this as a multi-label classification problem. Also, it is possible that the user query may have multiple intents and a multi-label model is a good choice to handle such a scenario.

For this reason, I have transformed the created a multi-label dataset, with the following 17 classes:
```
'ground_service', 'abbreviation', 'ground_fare', 'airline', 'city',
'aircraft', 'flight_no', 'airport', 'flight', 'quantity', 'meal',
'capacity', 'restriction', 'airfare', 'distance', 'flight_time', 'cheapest'
```

### Data
- `data` - The provided ATIS dataset
  - downlaod link: `https://drive.google.com/drive/folders/1I2cALZXOIaz9WnmdtubpVavflkUvFRHG?usp=share_link`
- `data_mlabel` - The multilabel version of provided ATIS dataset
  - download link: `https://drive.google.com/drive/folders/1-0VzuUa16j3nEcqywHVKlJONMPg0F40S?usp=share_link`
- `multilingual_data` - Parallel Multilingual translated data of provdied ATIS dataset
  - download link - `https://drive.google.com/drive/folders/1A-t73esVP27KbC9eAEMBv6klu8jpQSd-?usp=share_link`

### Modeling

The model architecture is a simple one, which I believe is a strong baseline for the task and can be used for handling multi-lingual queries.

1. Encoder

    As I am tring to train a multilingual model, the first step in the NLP pipeline would be to have a [`bert-base-multilingual-cased`](https://huggingface.co/bert-base-multilingual-cased) from huggingface-transformers. This BERT varient is a pretrained model on the top 104 languages with the largest Wikipedia using a masked language modeling (MLM) objective. We can easily extend our approach to handle queries in 104 languages, however the performance might differ between languages depending on the amount of data used in multi-lingual BERT pretraining.

2. Decoder (classifier)

    The decoder is a single linear layer mapping the encoder output to our 17-output classes.

3. Loss Function

    Binary Cross Entropy is a suitable loss function for multi-label modeling in this scenario.

#### Model

```


IntentClassifier(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(119547, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-11): 12 x BertLayer(
          (attention): BertAttention(
            (self): BertSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
          )
          (intermediate): BertIntermediate(
            (dense): Linear(in_features=768, out_features=3072, bias=True)
            (intermediate_act_fn): GELUActivation()
          )
          (output): BertOutput(
            (dense): Linear(in_features=3072, out_features=768, bias=True)
            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
        )
      )
    )
    (pooler): BertPooler(
      (dense): Linear(in_features=768, out_features=768, bias=True)
      (activation): Tanh()
    )
  )
  (classifier): Linear(in_features=768, out_features=17, bias=True)
  (criterion): BCEWithLogitsLoss()
)


```
#### Parameters
```
  | Name       | Type              | Params
-------------------------------------------------
0 | bert       | BertModel         | 177 M 
1 | classifier | Linear            | 13.1 K
2 | criterion  | BCEWithLogitsLoss | 0     
-------------------------------------------------
177 M     Trainable params
0         Non-trainable params
177 M     Total params
711.466   Total estimated model params size (MB)
```
# Instructions

## Setup

* Create a conda environment using:
    - `conda create -n "intent-clf-env" python=3.10.11`
* Activate the conda environment:
    - `conda activate intent-clf-env`
* Pylint Setup:
  * `conda install -c conda-forge pre_commit`
  * `pre-commit install`
  * `pre-commit autoupdate`
  * Note: Not keeping these dependencies in `requirements.txt` as they are only required for development purpose.

* Install dependencies
    - `pip install -r requirements.txt`
* In the repository you will find `.env.bkp` file. You need to create a copy of the file:
    - `cp .env.bkp .env`
    - Setup the environment variables:
        - ```
            PORT=8080
            CHECKPOINT_PATH="./model/best-checkpoint-v1.ckpt"
            ML_BINARIZER_PATH="./model/ml_binarizer.pkl"
          ```
* Download the model files in the `models` directory:
    - `best-checkpoint-v1.ckpt` - `https://drive.google.com/file/d/1-P-mIf9ChF04LzZ63EZiitUiGthVfYzm/view?usp=share_link`
    - `ml_binarizer.pkl` - `https://drive.google.com/file/d/1-Q5671xZmR54XSChXq9yv1yFN41-tZVJ/view?usp=share_link`
* Running the Flask server
    - `python server.py`

## Building or downloading docker image

### Building docker image
* The code has a `Dockerfile` pre-setup
* In the repository you will find `.env.bkp` file. You need to create a copy of the file:
    - `cp .env.bkp .env`
    - Setup the environment variables:
        - ```
            PORT=8080
            CHECKPOINT_PATH="./model/best-checkpoint-v1.ckpt"
            ML_BINARIZER_PATH="./model/ml_binarizer.pkl"
          ```
* Download the model files in the `models` directory:
    - `best-checkpoint-v1.ckpt` - `https://drive.google.com/file/d/1-P-mIf9ChF04LzZ63EZiitUiGthVfYzm/view?usp=share_link`
    - `ml_binarizer.pkl` - `https://drive.google.com/file/d/1-Q5671xZmR54XSChXq9yv1yFN41-tZVJ/view?usp=share_link`
* Build docker images:
    - `sudo docker build . -t lordzuko/intent-clf-service:v1.0.0`
* Running docker container
    - `sudo docker run -p 8080:8080 lordzuko/intent-clf-service:v1.0.0`

### Dowload pre-built docker image
* [DockerHub](https://hub.docker.com/repository/docker/lordzuko/intent-clf-service/)
* `docker pull lordzuko/intent-clf-service:v1.0.0`

### Limitation
* The Dockerfile currently does not support GPU usage even though the inference code works with GPU. This is because the chosen base image does not have NVIDIA-CUDA drivers installed.
# API Documentation
The documentation provides how to use the API, with `python`, `cURL` etc.
* POSTMAN documentation for the service can be found [here](https://documenter.getpostman.com/view/30635450/2s9YRB4CSg#6401b2aa-c4d3-4881-9ba6-7182af00ef43).

# Data Preparation and Model Training
- The process for model training and evaluation is described in notebook: `notebooks/multi_lingual_multilabel_intent_clf.ipynb`
- The process for model evalation during production scenario is described in notebook: `notebooks/Production_Evaluation.ipynb`
  - For running the `Production_Evaluation.ipynb` notebook, you can use the pre-computed predictions on train, test and valid split [here](https://drive.google.com/drive/folders/149twUWTkoHoBnv6-_Mbj0eqhtGPsuztq?usp=share_link)
  - The link to the collab notebook for `Production_Evaluation.ipynb` is [here](https://colab.research.google.com/drive/1AuuW0dw6oVPhBHGJlJ6mWCcjd0wsczKE?usp=sharing). This shows certain plots which are not shown in the committed notebook. 
# Evaluation


## Accuracy
```
Best Threshold: 0.30
Train Accuracy: 0.999
Val Accuracy: 0.998
Test Accuracy: 0.995
```
## AU-ROC
### Overall
```
Best Threshold: 0.20
Train AUROC: 0.874
Val AUROC: 0.833
Test AUROC: 0.805
```
### Per Intent Class

You can format the validation and test results for markdown as follows:

| Label            | Validation AUROC | Test AUROC     |
|------------------|------------------|-----------------|
| abbreviation     | 1.000000         | 0.999959        |
| aircraft         | 0.998834         | 0.997295        |
| airfare          | 0.999978         | 0.989687        |
| airline          | 0.999845         | 0.992285        |
| airport          | 1.000000         | 0.999959        |
| capacity         | 1.000000         | 0.999260        |
| cheapest         | 0.000000         | 0.000000        |
| city             | 1.000000         | 0.936331        |
| distance         | 1.000000         | 1.000000        |
| flight           | 0.999852         | 0.989004        |
| flight_no        | 0.992795         | 1.000000        |
| flight_time      | 0.992006         | 1.000000        |
| ground_fare      | 1.000000         | 0.997722        |
| ground_service   | 0.998485         | 0.999989        |
| meal             | 0.992806         | 0.926189        |
| quantity         | 0.999711         | 0.996064        |
| restriction      | 0.000000         | 0.000000        |


## Classificaiton Report

| Label         | Train Support | Validation Support | Test Support | Validation Precision | Test Precision | Validation Recall | Test Recall | Validation F1-Score | Test F1-Score |
|---------------|---------------|--------------------|--------------|-----------------------|-----------------|------------------|------------|-------------------|---------------|
| abbreviation  | 309.0         | 15.0               | 78.0         | 1.00                | 0.97           | 1.00             | 0.99       | 1.00              | 0.98          |
| aircraft      | 227.0         | 10.0               | 24.0         | 0.83                | 0.88           | 1.00             | 0.88       | 0.91              | 0.88          |
| airfare       | 1190.0        | 73.0               | 183.0        | 0.97                | 0.97           | 1.00             | 0.93       | 0.99              | 0.95          |
| airline       | 421.0         | 29.0               | 90.0         | 1.00                | 0.96           | 0.97             | 0.96       | 0.98              | 0.96          |
| airport       | 53.0          | 1.0                | 39.0         | 1.00                | 0.95           | 1.00             | 0.97       | 1.00              | 0.96          |
| capacity      | 46.0          | 2.0                | 63.0         | 1.00                | 1.00           | 1.00             | 0.94       | 1.00              | 0.97          |
| cheapest      | 3.0           | 0.0                | 0.0          | 0.00                | 0.00           | 0.00             | 0.00       | 0.00              | 0.00          |
| city          | 51.0          | 3.0                | 15.0         | 1.00                | 1.00           | 1.00             | 0.40       | 1.00              | 0.57          |
| distance      | 58.0          | 2.0                | 30.0         | 1.00                | 0.39           | 1.00             | 1.00       | 1.00              | 0.57          |
| flight        | 9822.0        | 510.0              | 1881.0       | 1.00                | 0.99           | 0.99             | 0.98       | 1.00              | 0.99          |
| flight_no     | 43.0          | 2.0                | 27.0         | 1.00                | 1.00           | 0.50             | 1.00       | 0.67              | 1.00          |
| flight_time   | 151.0         | 8.0                | 3.0          | 0.78                | 0.27           | 0.88             | 1.00       | 0.82              | 0.43          |
| ground_fare   | 51.0          | 3.0                | 21.0         | 1.00                | 0.29           | 1.00             | 0.95       | 1.00              | 0.44          |
| ground_service| 672.0         | 36.0               | 108.0        | 0.97                | 0.92           | 1.00             | 1.00       | 0.99              | 0.96          |
| meal          | 17.0          | 1.0                | 18.0         | 0.00                | 0.00           | 0.00             | 0.00       | 0.00              | 0.00          |
| quantity      | 142.0         | 5.0                | 9.0          | 0.83                | 0.21           | 1.00             | 1.00       | 0.91              | 0.35          |
| restriction   | 15.0          | 0.0                | 0.0          | 0.00                | 0.00           | 0.00             | 0.00       | 0.00              | 0.00          |


| Summary         | Train Support | Validation Support | Test Support | Validation Precision | Test Precision | Validation Recall | Test Recall | Validation F1-Score | Test F1-Score |
|---------------|---------------|--------------------|--------------|-----------------------|-----------------|------------------|------------|-------------------|---------------|
| micro avg     | 13271.0       | 700.0              | 2589.0       | 0.99                | 0.93           | 0.99             | 0.97       | 0.99              | 0.95          |
| macro avg     | 13271.0       | 700.0              | 2589.0       | 0.79                | 0.64           | 0.78             | 0.76       | 0.78              | 0.65          |
| weighted avg  | 13271.0       | 700.0              | 2589.0       | 0.99                | 0.96           | 0.99             | 0.97       | 0.99              | 0.96          |
| samples avg   | 13271.0       | 700.0              | 2589.0       | 0.99                | 0.95           | 0.99             | 0.97       | 0.99              | 0.96          |




# Future Improvements
1. The model is currently being loaded from checkpoint, which has optimizer and gradient placeholders, which is increasing the file size. A better way would be to save model dict and load the model from state dict. This will reduce the model file size, which will ultimately decrease the size of resulting docker image.
2. Evaluation is not done at language level, which can be important to make any language specific improvements and updates.
3. The `/intent` api endpoint does not provide an `api_version` field which could be important to keep track of the current version of intent classification service which is producing the output. The downstream applications might take advantage of it to handle business logic or update them according to api_version.
   1. It can also be important in AB-testing
   2. Also, data management could take advantage of this field
4. The `/intent` api does not check for the language. This could be problametic as our model currently only supports for 3 languages, whereas the tokenizer we are using can support 104 langauges. Not detecting the supported languages could lead to unforseen model performance issues.
5. Data imbalance among intent classes is currently a bottlenect and upsampling needs to be down for low data classes. We can use translation and paraphrasing to tackle the data imbalance issues. 
6. Update Dockerfile to support GPU usage.
7. `NannyML` installation has numpy version mismatch with the rest of the code, so it might create issues with building Docker image and has not been tested yet. However, given that it is not required during inference, I suggest install it in separate environment for testing for now.


CHANGELOG.md
# **Change Log** 📜📝

All notable changes to the "**Intent Classification Service**" WhatItIs/program/extension/API/whatever will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

---
## [v1.0.0] - 2023-10-21

### Added

* Added BERT model - `bert-base-multilingual-cased` from huggingface as multilingual encoder model
* Added error handling codes for /intent endpoint as per the following specifications:
  * BODY_MISSING, 400
  * TEXT_MISSING, 400
  * INVALID_TYPE, 400
  * TEXT_EMPTY, 400
  * INTERNAL_ERROR, 500
* Added model training and production evaluation notebook in `notebooks/`
* Added Docker files
* Added POSTMAN api documentation
### Changed

* Modifed the /ready logic with simple `model` is `None` check as the `IntentClassifier` does not have any `ready()` method.

### Fixed

* In the initial code the model was loaded after flask server was up. Moved the model loading before starting server fixed the issue.


server.py
from __future__ import annotations

import logging

import torch
from flask import Flask
from flask import jsonify
from flask import request

from config import LABEL_COLUMNS
from config import MAX_TOKEN_COUNT
from config import PORT
from src.utils import get_model_and_tokenizer

app = Flask(__name__)
MODEL, TOKENIZER = get_model_and_tokenizer(eval=True)
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

@app.route("/ready")
def ready():
    if MODEL:
        return "OK", 200
    else:
        return "Not ready", 423


@app.route("/intent", methods=["POST"])
def intent():
    try:
        request_data = request.get_json()

        if not request_data:
            return (
                jsonify(label="BODY_MISSING", message="Request doesn't have a body."),
                400,
            )

        if "text" not in request_data:
            return (
                jsonify(
                    label="TEXT_MISSING",
                    message='"text" missing from request body.',
                ),
                400,
            )

        text = request_data["text"]

        if not isinstance(text, str):
            return jsonify(label="INVALID_TYPE", message='"text" is not a string.'), 400

        if not text:
            return jsonify(label="TEXT_EMPTY", message='"text" is empty.'), 400

        encoding = TOKENIZER.encode_plus(
            text,
            add_special_tokens=True,
            max_length=MAX_TOKEN_COUNT,
            return_token_type_ids=False,
            padding="max_length",
            return_attention_mask=True,
            return_tensors="pt",
        )
        
        encoding.to(device)
        _, test_prediction = MODEL(encoding["input_ids"], encoding["attention_mask"])
        test_prediction = test_prediction.flatten()
        test_logits = torch.sigmoid(test_prediction).detach().cpu()

        values = torch.topk(test_logits, 3).values
        indices = torch.topk(test_logits, 3).indices

        pred_intents = []
        for i, v in zip(indices, values):
            pred_intents.append(
                {"label": LABEL_COLUMNS[i], "confidence": round(v.item(), 3)},
            )
        return jsonify({"intents": pred_intents}), 200

    except Exception:
        error_message = "An error occurred while processing the request."
        logging.exception("Exception in /intent endpoint:")
        return jsonify(label="INTERNAL_ERROR", message=error_message), 500


def main():
    app.run(host="0.0.0.0", port=PORT)


if __name__ == "__main__":
    main()


src/utils.py
from __future__ import annotations

import torch
from transformers import BertTokenizerFast

from config import CHECKPOINT_PATH
from config import LABEL_COLUMNS
from config import ML_BERT_MODEL_NAME
from src.intent_classifier import IntentClassifier


def get_model_and_tokenizer(eval=True):
    """Get the model and tokenizer for usage

    Args:
        eval (bool, optional): Set model in evaluation mode. Defaults to True.

    Returns:
        model, tokenizer: return the loaded model and tokenizer
    """
    tokenizer = BertTokenizerFast.from_pretrained(ML_BERT_MODEL_NAME)
    model = IntentClassifier.load_from_checkpoint(
        CHECKPOINT_PATH,
        n_classes=len(LABEL_COLUMNS),
    )
    if eval:
        model.eval()
        model.freeze()
    if torch.cuda.is_available():
        model.to("cuda")

    return model, tokenizer


src/data_loaders.py
from __future__ import annotations

import pandas as pd
import pytorch_lightning as pl
import torch
from torch.utils.data import DataLoader
from torch.utils.data import Dataset
from transformers import BertTokenizerFast

from config import LABEL_COLUMNS


class IntentClassificationDataset(Dataset):
    def __init__(
        self,
        data: pd.DataFrame,
        tokenizer: BertTokenizerFast,
        max_token_len: int = 100,
    ):
        super().__init__()
        self.tokenizer = tokenizer
        self.data = data
        self.max_token_len = max_token_len

    def __len__(self):
        return len(self.data)

    def __getitem__(self, index: int):
        data_row = self.data.iloc[index]

        text = data_row.text
        lang = data_row.lang
        labels = data_row[LABEL_COLUMNS]

        encoding = self.tokenizer(
            text,
            add_special_tokens=True,
            max_length=self.max_token_len,
            return_token_type_ids=False,
            padding="max_length",
            truncation=True,
            return_attention_mask=True,
            return_tensors="pt",
        )

        return dict(
            text=text,
            lang=lang,
            input_ids=encoding["input_ids"].flatten(),
            attention_mask=encoding["attention_mask"].flatten(),
            labels=torch.FloatTensor(labels),
        )


class IntentClassificationDataModule(pl.LightningDataModule):
    def __init__(
        self,
        train_df: pd.DataFrame,
        valid_df: pd.DataFrame,
        test_df: pd.DataFrame,
        tokenizer: BertTokenizerFast,
        batch_size=8,
        max_token_len=100,
    ):
        super().__init__()
        self.batch_size = batch_size
        self.train_df = train_df
        self.valid_df = valid_df
        self.test_df = test_df
        self.tokenizer = tokenizer
        self.max_token_len = max_token_len

    def setup(self, stage=None):
        self.train_dataset = IntentClassificationDataset(
            self.train_df,
            self.tokenizer,
            self.max_token_len,
        )

        self.val_dataset = IntentClassificationDataset(
            self.valid_df,
            self.tokenizer,
            self.max_token_len,
        )

        self.test_dataset = IntentClassificationDataset(
            self.test_df,
            self.tokenizer,
            self.max_token_len,
        )

    def train_dataloader(self):
        return DataLoader(
            self.train_dataset,
            batch_size=self.batch_size,
            shuffle=True,
            num_workers=2,
        )

    def val_dataloader(self):
        return DataLoader(self.val_dataset, batch_size=self.batch_size, num_workers=2)

    def test_dataloader(self):
        return DataLoader(self.test_dataset, batch_size=self.batch_size, num_workers=2)


src/intent_classifier.py
from __future__ import annotations

import pytorch_lightning as pl
import torch
import torch.nn as nn
from torchmetrics.functional.classification import auroc
from transformers import AdamW
from transformers import BertModel
from transformers import get_linear_schedule_with_warmup

from config import LABEL_COLUMNS
from config import ML_BERT_MODEL_NAME


class IntentClassifier(pl.LightningModule):
    """
    Intent Classifier Model Class
    """

    def __init__(self, n_classes: int, n_training_steps=None, n_warmup_steps=None):
        super().__init__()
        self.bert = BertModel.from_pretrained(ML_BERT_MODEL_NAME, return_dict=True)
        self.classifier = nn.Linear(self.bert.config.hidden_size, n_classes)
        self.n_training_steps = n_training_steps
        self.n_warmup_steps = n_warmup_steps
        self.criterion = nn.BCEWithLogitsLoss()
        self.training_step_outputs: list[dict] = list()

    def forward(self, input_ids, attention_mask, labels=None):
        output = self.bert(input_ids, attention_mask=attention_mask)
        output = self.classifier(output.pooler_output)

        loss = 0
        if labels is not None:
            loss = self.criterion(output, labels)
        return loss, output

    def training_step(self, batch, batch_idx):
        input_ids = batch["input_ids"]
        attention_mask = batch["attention_mask"]
        labels = batch["labels"]
        loss, outputs = self(input_ids, attention_mask, labels)
        self.training_step_outputs.append(
            {"loss": loss, "predictions": outputs, "labels": labels},
        )
        self.log("train_loss", loss, prog_bar=True, logger=True)
        return {"loss": loss, "predictions": outputs, "labels": labels}

    def validation_step(self, batch, batch_idx):
        input_ids = batch["input_ids"]
        attention_mask = batch["attention_mask"]
        labels = batch["labels"]
        loss, outputs = self(input_ids, attention_mask, labels)
        self.log("val_loss", loss, prog_bar=True, logger=True)
        return loss

    def test_step(self, batch, batch_idx):
        input_ids = batch["input_ids"]
        attention_mask = batch["attention_mask"]
        labels = batch["labels"]
        loss, outputs = self(input_ids, attention_mask, labels)
        self.log("test_loss", loss, prog_bar=True, logger=True)
        return loss

    def on_train_epoch_end(self):
        labels = []
        predictions = []
        for output in self.training_step_outputs:
            for out_labels in output["labels"].detach().cpu():
                labels.append(out_labels)
            for out_predictions in output["predictions"].detach().cpu():
                predictions.append(out_predictions)

        labels = torch.stack(labels).int()
        predictions = torch.stack(predictions)

        for i, name in enumerate(LABEL_COLUMNS):
            class_roc_auc = auroc(predictions[:, i], labels[:, i], task="binary")
            self.logger.experiment.add_scalar(
                f"{name}_roc_auc/Train",
                class_roc_auc,
                self.current_epoch,
            )

        self.training_step_outputs.clear()

    def configure_optimizers(self):
        optimizer = AdamW(self.parameters(), lr=2e-5)

        scheduler = get_linear_schedule_with_warmup(
            optimizer,
            num_warmup_steps=self.n_warmup_steps,
            num_training_steps=self.n_training_steps,
        )

        return dict(
            optimizer=optimizer,
            lr_scheduler=dict(scheduler=scheduler, interval="step"),
        )


config.py
from __future__ import annotations

import logging
import os

import joblib
from dotenv import load_dotenv

# Configure logging
logging.basicConfig(level=logging.INFO)

BASE_DIR = os.path.abspath(os.path.dirname(__file__))
ENV_PATH = os.path.join(BASE_DIR, ".env")
print(os.path.abspath(ENV_PATH))
load_dotenv(dotenv_path=ENV_PATH)

API_VERSION = "v1.0.0"
RANDOM_SEED = 42
MAX_TOKEN_COUNT = 100
ML_BERT_MODEL_NAME = "bert-base-multilingual-cased"


EXPECTED_LABELS = [
    "abbreviation",
    "aircraft",
    "airfare",
    "airline",
    "airport",
    "capacity",
    "cheapest",
    "city",
    "distance",
    "flight",
    "flight_no",
    "flight_time",
    "ground_fare",
    "ground_service",
    "meal",
    "quantity",
    "restriction",
]
# LABELWISE_THRESHOLDS = [
#     0.60,
#     0.60,
#     0.2,
#     0.30,
#     0.30,
#     0.2,
#     0.2,
#     0.30,
#     0.1,
#     0.2,
#     0.1,
#     0.2,
#     0.1,
#     0.30,
#     0.2,
#     0.30,
#     0.2,
# ]

PORT = int(os.environ["PORT"])
CHECKPOINT_PATH = os.environ["CHECKPOINT_PATH"]
LABEL_COLUMNS = joblib.load(os.environ["ML_BINARIZER_PATH"]).classes_
assert (
    list(LABEL_COLUMNS) == EXPECTED_LABELS
), "The order of labels are changed, please ensure they match."
# assert len(LABELWISE_THRESHOLDS) == len(
#     EXPECTED_LABELS
# ), "The count of LABELWISE_TRHESHOLDS should be equal to EXPECTED_LABELS"
